app/javascript/react/screens/App/Overview/screens/MappingWizard/components/MappingWizardDatastoresStep/components/DatastoresStepForm.js

```diff
@@ -49,28 +52,93 @@ class DatastoresStepForm extends React.Component {
   }
 
   selectSourceDatastore(sourceDatastore) {
-    this.setState(prevState => {
-      const isAlreadySelected = prevState.selectedSourceDatastores.some(
-        datastore => datastore.id === sourceDatastore.id
+    const { selectedTargetDatastore, selectedSourceDatastores } = this.state;
+    const { value: datastoresStepMappings } = this.props.input;
+    const { showAlertAction, hideAlertAction } = this.props;
+
+    const mappingExistsForTargetDatastore =
+      selectedTargetDatastore &&
+      datastoresStepMappings.some(targetClusterWithDatastoreMappings =>
+        targetClusterWithDatastoreMappings.nodes.some(
+          targetDatastoreWithSourceDatastores =>
+            targetDatastoreWithSourceDatastores.id ===
+            selectedTargetDatastore.id
+        )
       );
-      if (isAlreadySelected) {
+
+    const isNotAlreadySelected = !selectedSourceDatastores.some(datastore => {
+      return datastore.id === sourceDatastore.id;
+    });
+
+    if (
```

Comment:
This block of code took me a few read-throughs to get-- could benefit from some comments later. Looks good though, let's keep moving.